Sources/FoundationEssentials/Data/Data+Writing.swift

```diff
@@ -387,44 +389,112 @@ private func writeToFileAux(path inPath: PathOrURL, buffer: UnsafeRawBufferPoint
+            let cbSize = wcslen(pwszPath) * MemoryLayout<WCHAR>.size
+            let dwSize = DWORD(MemoryLayout<FILE_RENAME_INFO>.size + cbSize)
+            withUnsafeTemporaryAllocation(byteCount: Int(dwSize),
+                                          alignment: MemoryLayout<FILE_RENAME_INFO>.alignment) { pBuffer in
+                var pInfo = pBuffer.baseAddress?.bindMemory(to: FILE_RENAME_INFO.self, capacity: 1)
+                pInfo?.pointee.Flags = FILE_RENAME_FLAG_POSIX_SEMANTICS | FILE_RENAME_FLAG_REPLACE_IF_EXISTS
+                pInfo?.pointee.RootDirectory = nil
+                pInfo?.pointee.FileNameLength = DWORD(cbSize)
+                withUnsafePointer(to: &pInfo?.pointee.FileName) { pFileName in
+                    wcscpy_s(pFileName, wcslen(pwszPath) + 1, pwszPath)
+                }
 
-        _close(fd)
-        fd = -1
+                if !SetFileInformationByHandle(hFile, FileRenameInfoEx, pInfo, dwSize) {
+                    let dwError = GetLastError()
+                    if dwError == ERROR_ACCESS_DENIED {
+                        let dwAttributes = GetFileAttributesW(pwszPath)
+                        if dwAttributes > 0,
+                                dwAttributes & (FILE_ATTRIBUTE_READONLY | FILE_ATTRIBUTE_HIDDEN | FILE_ATTRIBUTE_SYSTEM) != 0 {
+                            // The target file is read-only, hidden or a system file. Remove those attributes and try again.
+                            if SetFileAttributesW(pwszPath, dwAttributes & ~(FILE_ATTRIBUTE_READONLY | FILE_ATTRIBUTE_HIDDEN | FILE_ATTRIBUTE_SYSTEM)) {
+                                if SetFileInformationByHandle(hFile, FileRenameInfoEx, pInfo, dwSize) {
+                                    return
+                                }
+                                // Try to restore the attributes, but ignore any error
+                                _ = SetFileAttributesW(pwszPath, dwAttributes)
+                            }
+                        }
+                    }
+                    guard dwError == ERROR_NOT_SAME_DEVICE else {
+                        _ = DeleteFileW(pwszAuxiliaryPath)
+                        throw CocoaError.errorWithFilePath(inPath, win32: dwError, reading: false)
+                    }
 
-        try auxPath.withNTPathRepresentation { pwszAuxiliaryPath in
-            guard MoveFileExW(pwszAuxiliaryPath, pwszPath, MOVEFILE_COPY_ALLOWED | MOVEFILE_REPLACE_EXISTING | MOVEFILE_WRITE_THROUGH) else {
-                let dwError = GetLastError()
-                _ = DeleteFileW(pwszAuxiliaryPath)
-                throw CocoaError.errorWithFilePath(inPath, win32: dwError, reading: false)
+                    _ = CloseHandle(hFile)
+                    hFile = INVALID_HANDLE_VALUE
+
+                    // The move is across volumes.
+                    guard MoveFileExW(pwszAuxiliaryPath, pwszPath, MOVEFILE_COPY_ALLOWED | MOVEFILE_REPLACE_EXISTING) else {
```

Comment:
The previous version of this call also set `MOVEFILE_WRITE_THROUGH`, should we preserve that here?

Reply:
Good question. The Linux/Darwin equivalent that we use as `rename` so ideally we'd opt for whatever behavior is most similar to `rename`. I don't know offhand what the answer is there (I didn't find anything particularly specific in the man page) but I can try to find out what the semantics are of `rename` to match here